File: solutions/code-carl/ch01_array/T6_LC0059_GenerateMatrix.java
/**
 * Practice template for:
 * - LeetCode 59. Spiral Matrix II
 *   https://leetcode.cn/problems/spiral-matrix-ii/
 */
public class T6_LC0059_GenerateMatrix {
    public int[][] generateMatrix(int n) {
        int[][] matrix = new int[n][n];
        int round = n / 2;
        int num = 1;
        for (int r = 0; r < round; r++) {
            for (int j = r; j < n - r - 1; j++) matrix[r][j] = num++;
            for (int i = r; i < n - r - 1; i++) matrix[i][n - r - 1] = num++;
            for (int j = n - r - 1; j > r; j--) matrix[n - r - 1][j] = num++;
            for (int i = n - r - 1; i > r; i--) matrix[i][r] = num++;
        }
        if (n % 2 == 1) {
            matrix[n / 2][n / 2] = n * n;
        }
        return matrix;
    }
}

File: solutions/code-carl/ch01_array/t6_lc0059_generate_matrix.py
"""LeetCode 59. Spiral Matrix II
https://leetcode.cn/problems/spiral-matrix-ii/
"""
from itertools import count
from typing import List


class Solution:
    def generateMatrix(self, n: int) -> List[List[int]]:
        res: List[List[int]] = [[0] * n for _ in range(n)]
        cnt = count(1)
        for r in range(n // 2):
            for j in range(r, n - r - 1):
                res[r][j] = next(cnt)
            for i in range(r, n - r - 1):
                res[i][n - r - 1] = next(cnt)
            for j in range(n - r - 1, r, -1):
                res[n - r - 1][j] = next(cnt)
            for i in range(n - r - 1, r, - 1):
                res[i][r] = next(cnt)
        if n % 2 == 1:
            res[n // 2][n // 2] = next(cnt)
        return res


if __name__ == "__main__":
    solution = Solution()

    assert solution.generateMatrix(1) == [[1]]
    assert solution.generateMatrix(3) == [[1, 2, 3], [8, 9, 4], [7, 6, 5]]
    print("All tests passed.")
